File: src/main/scala/com/microsoft/azure/cosmosdb/kafka/connect/sink/CosmosDBSinkConnector.scala
package com.microsoft.azure.cosmosdb.kafka.connect.sink

import java.util

import com.microsoft.azure.cosmosdb.kafka.connect.common.ErrorHandler.HandleRetriableError
import com.microsoft.azure.cosmosdb.kafka.connect.config.{ConnectorConfig, CosmosDBConfig}
import org.apache.kafka.common.config.ConfigDef
import org.apache.kafka.connect.connector.Task
import org.apache.kafka.connect.sink.SinkConnector

import scala.collection.JavaConverters._
import scala.util.{Failure, Success}

class CosmosDBSinkConnector extends SinkConnector with HandleRetriableError {


  private var configProps: util.Map[String, String] = _


  override def version(): String = getClass.getPackage.getImplementationVersion

  override def start(props: util.Map[String, String]): Unit = {
    logger.info("Starting CosmosDBSinkConnector")

    try {
      initializeErrorHandler(props.get(org.apache.kafka.connect.runtime.ConnectorConfig.ERRORS_RETRY_TIMEOUT_CONFIG).toInt) // TODO: test

      val config = CosmosDBConfig(ConnectorConfig.sinkConfigDef, props)
      HandleRetriableError(Success(config))
    }
    catch{
      case f: Throwable =>
        logger.error(s"Couldn't start Cosmos DB Sink due to configuration error: ${f.getMessage}", f)
        HandleRetriableError(Failure(f))
    }

    configProps = props

  }

  override def stop(): Unit = {
    logger.info("Stopping CosmosDBSinkConnector")
  }

  override def taskClass(): Class[_ <: Task] = classOf[CosmosDBSinkTask]

  override def taskConfigs(maxTasks: Int): util.List[util.Map[String, String]] = {
    logger.info(s"Setting task configurations for $maxTasks workers with properties $this.configProps")
    println(this.configProps)

    (1 to maxTasks).map(_ => this.configProps).toList.asJava

  }
  override def config(): ConfigDef = ConnectorConfig.sinkConfigDef

}

File: src/main/java/com/microsoft/azure/cosmosdb/kafka/connect/sink/CosmosDBSinkConnector.java
package com.microsoft.azure.cosmosdb.kafka.connect.sink;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import org.apache.kafka.common.config.ConfigDef;
import org.apache.kafka.connect.connector.Task;
import org.apache.kafka.connect.sink.SinkConnector;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * A Sink connector that publishes topic messages to CosmosDB.
 */
public class CosmosDBSinkConnector extends SinkConnector {

    private static final Logger logger = LoggerFactory.getLogger(CosmosDBSinkConnector.class);
    private Map<String, String> configProps;

    @Override
    public void start(Map<String, String> props) {
        logger.debug("Starting CosmosDB sink connector.");
        configProps = props;
    }

    @Override
    public Class<? extends Task> taskClass() {
        return CosmosDBSinkTask.class;
    }

    @Override
    public List<Map<String, String>> taskConfigs(int maxTasks) {
        logger.info("Setting task configurations for {} workers.", maxTasks);
        final List<Map<String, String>> configs = new ArrayList<>(maxTasks);
        for (int i = 0; i < maxTasks; ++i) {
          configs.add(configProps);
        }
        return configs;
      }

    @Override
    public void stop() {
        logger.debug("Stopping CosmosDB sink connector.");
    }

    @Override
    public ConfigDef config() {
        logger.debug("config invoked");

        ConfigDef configDef = new ConfigDef();
        new SinkSettings().getAllSettings().stream().forEachOrdered(setting -> setting.toConfigDef(configDef));

        logger.debug("Sink ConfigDef with {} settings", configDef.configKeys().size()
        );

        return configDef;
    }
    
    @Override
    public String version() {
        String val = this.getClass().getPackage().getImplementationVersion();
        logger.debug("version {}", val);
        return val;
    }

}
